lib/core/remove-file.ts

```diff
@@ -0,0 +1,70 @@
+import fs from "fs";
+import path from "path";
+import glob from "glob";
+
+import { alerts } from "./alerts";
+import {
+  getScssPathFromTypeDefinition,
+  getTypeDefinitionPath
+} from "../typescript";
+
+/**
+ * Given a single file remove the generated types if they exist
+ *
+ * @param file the SCSS file to generate types for
+ *
+ */
+
+export const removeFile = (file: string): void => {
+  if (fs.existsSync(file)) {
+    return;
+  }
+  const path = getTypeDefinitionPath(file);
+  try {
+    if (fs.existsSync(path)) {
+      fs.unlink(path, err => {
+        if (err) {
+          alerts.error(`Could not remove ${path}`);
+        } else {
+          alerts.success(`[REMOVED TYPES] ${path}`);
+        }
+      });
+    }
+  } catch (error) {
+    alerts.notice(`[NO GENERATED TYPES] for ${file}`);
+  }
+};
+/**
+ * Gets the directory names of the files that are passed in.
+ *
+ * @param files the files to obtain the name of the directories for.
+ *
+ * @returns string[]   an array of directory names from the files.
+ */
+const getDirectoriesOfFiles = (files: string[]): string[] => {
+  const directories = files.map(file => path.dirname(file));
+  return directories.filter((dir, index) => directories.indexOf(dir) === index);
+};
+
+/**
+ * Removes the typed files of those scss files that have been deleted or moved.
+ *
+ * @param files the paths of files to search through to remove typed files
+ *
+ */
+export const removeUnusedTypeFiles = (files: string[]): void => {
```

Comment:
Would you mind explaining how this function works? I see we are looping over files and doing a directory lookup, then doing a glob in each of these directories, and transforming back and forth. Does that mean we are potentially double processing directories and hitting the filesystem 3+ times per SCSS file?

Can we instead take the `files` that match the glob in `generate`, then do an equivalent glob for `*.d.ts` and do a subtraction. 

```
[*.d.ts files that match the glob] - [*.scss files that match the glob] = [files to remove]
```

That way, we'd only be doing one glob rather than `n` multiple times? I think it would be simpler to comprehend and more performant.

Reply:
Before my change, the current state of the world is that a user passes a pattern to the cli in order for the cli to include files that adhere to the pattern. Once we found those files we can either update or add new type definitions. Knowing this state of the world, the challenge is how to figure out that a `*.scss` file has been removed if a user is not running the cli with the `--watch` flag. 

The problem with the current state of the world is that a deleted `*.scss` file is removed so our pattern matching will not return to us or tell us of its existence. So what this function tries to accomplish is use the files the user has included in order to find obsolete type definitions. In this way we avoid being super destructive such as delete files that the user has excluded. 

The algorithm I decided to use (in order to interfere the least) is basically given a file, get the directory, filter out any duplicate directories that we have seen (in order to have unique directories), given a directory (or directories) find all type definition files within it, remove all typed definitions that do not have a corresponding *.scss file. 

A diagram for future reference (just in case the code doesn't make sense or to help it evolve):
![Untitled Diagram](https://user-images.githubusercontent.com/10718803/95545768-a19aba00-09b3-11eb-850b-72dc6401caa2.png)

Reply:
What would happen if we have a file, say `root/example/style.scss`, and we run the generator to produce `root/example/style.scss.d.ts`. Then  `root/example/style.scss` is removed, and the generator is run again. Now that there is no file that contains the directory `root/example`, so we wouldn't look in this directory any longer and we'd still leave the type definition file there?

My concern with this general approach is that if it's not 100% accurate, it could cause more pain/confusion for people using this tool as well as maintaining since there will be continuous edge cases. 

This was my motivation for instead trying to reuse the glob passed in. For example, if they pass `./src/**/*.scss` could we instead find a way to convert that to `./src/**/*.scss.d.ts`? I can imagine the challenge with this approach would be all of the different possible globs that can be passed so this also may not be a viable option.

Reply:
That's a fair point. I think in that particular situation we would have to carry some kind of state from one call to another which just seems wrong to me since we don't know the frequency of the generator being run to keep an accurate state but more importantly that the tool should keep state. So that situation would break the whole logic and accuracy.

That for sure is a concern.

That could be an option and can only look at the pattern within the pattern they passed. So we can say that we will adhere to that and will not look at any other files that are not part of the pattern. In addition we can say as long as they glob pattern adheres to the `*.scss` ending for us to get the correct typed file name.

If you think this is feature that can be pushed off a bit more I could remove this second commit and only release the first commit which is the remove files while using the `--watch` flag.

Reply:
Yea, let's start with only while the `--watch` is running 👍 

This solution is clean/robust and I imagine this could be enough for most people requesting this feature?

I don't think we need to make it configurable either. I can't imagine a use case where someone removes a `*.scss` file while using this tool and they _don't_ want it removed. If there are extreme edge cases that arise after this, we can consider making an option to disable it...

Reply:
ok sounds good. I'll remove the second commit and keep the first.